iib/web/api_v1.py

```diff
@@ -168,24 +192,43 @@ def get_build_logs(request_id):
     :rtype: flask.Response
     :raise NotFound: if the request is not found or there are no logs for the request
     :raise Gone: if the logs for the build request have been removed due to expiration
+    :raise ValidationError: if the request has not completed yet
     """
     request_log_dir = flask.current_app.config['IIB_REQUEST_LOGS_DIR']
-    if not request_log_dir:
+    s3_bucket_name = flask.current_app.config['IIB_AWS_S3_BUCKET_NAME']
+    if not s3_bucket_name and not request_log_dir:
         raise NotFound()
 
     request = Request.query.get_or_404(request_id)
-    log_file_path = os.path.join(request_log_dir, f'{request_id}.log')
-    if not os.path.exists(log_file_path):
+
+    finalized = request.state.state_name in RequestStateMapping.get_final_states()
+    if not finalized:
+        raise ValidationError(
+            f'The request {request_id} is not complete yet.'
+            ' logs will be available once the request is complete.'
+        )
+
+    # If S3 bucket is configured, fetch the log file from the S3 bucket.
+    # Else, check if logs are stored on the system itself and return them.
+    # Otherwise, raise an IIBError.
+    if s3_bucket_name:
+        log_file = _get_artifact_file_from_s3_bucket(
+            'request_logs', f'{request_id}.log', request, s3_bucket_name
+        )
+        return flask.Response(log_file.read(), mimetype='text/plain')
+
+    local_log_file_path = os.path.join(request_log_dir, f'{request_id}.log')
+    if not os.path.exists(local_log_file_path):
         expired = request.temporary_data_expiration < datetime.utcnow()
         if expired:
             raise Gone(f'The logs for the build request {request_id} no longer exist')
-        finalized = request.state.state_name in RequestStateMapping.get_final_states()
-        if finalized:
-            raise NotFound()
-        # The request may not have been initiated yet. Return empty logs until it's processed.
-        return flask.Response('', mimetype='text/plain')
+        raise IIBError(
+            'IIB is done processing the request and could not find logs.'
+            ' Please make sure either an S3 bucket is configured or the logs are'
+            ' stored locally in a directory by specifying IIB_REQUEST_LOGS_DIR'
```

Comment:
I don't think it makes sense to mention `IIB_REQUEST_LOGS_DIR` to users. Maybe the raised error should be just the first sentence. We can log a warning with more information that would help the IIB maintainers debug issues.

Reply:
@yashvardhannanavati Do you agree with Luiz proposal?

Reply:
So, we have a similar error message in `related_bundles` which is why I decided to keep it consistent here. But if you feel otherwise strongly, I can change it :)

Reply:
Apologies for not replying to this earlier

Reply:
Changed it :)

Reply:
Thanks @yashvardhannanavati !